### PR TITLE
Networking revamp

### PR DESCRIPTION
**Pull request by luxaritas**
* Rewrote PageDataMixin as FetchMixin, acting more like Nuxt to resolve caching issues and generally be cleaner
* Overhauled network types
* Refactored homepage
* Refactored notifications
* Fixed a number of smaller misc errors

### CODE REVIEW

File: src/store/FetchData.vuex.ts [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+const VuexModule = createModule({
+  strict: false,
+});
+
+interface CacheData {data: any};
+interface CacheError {error: Error};
+
+export function isCacheData(cache: CacheData | CacheError): cache is CacheData {
+  return !!(cache && (cache as CacheData).data);
+}
```
Comment (guyguy2001) [inline src/store/FetchData.vuex.ts:14]: Couldn't data be `false` or `0`? If not maybe you can change the interface to `{data: object}`

File: src/views/feed/ActivityFeed/components/CommentMessageItem.vue [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+  import Utils from '@/utils/utils';
+  import MessageItem from './MessageItem.vue';
+
+  @Component({
+    components: { MessageItem },
+  })
+  export default class CommentMessageItem extends Vue {
+    @Prop() notification!: CommentNotificationItem;
+    
+    @Prop() message!: CommentNotificationMessage;
```
Comment (guyguy2001) [inline src/views/feed/ActivityFeed/components/CommentMessageItem.vue:30]: `{required:true}`

File: src/views/feed/ActivityFeed/components/MessageItem.vue [hunk 3]
```diff
@@ -15,17 +15,12 @@
...
-            <a :href="`/${content.node.node_type}s/${content.node.id}`">
-              {{ content.node.title }}</a
-            >
-          </template>
+        <p style="margin-top:5px">
+          <slot/>
         </p>
       </div>
-      <div v-dompurify-html="content.body || content" style="word-wrap: break-word;" />
+      <!--div v-dompurify-html="content.body || content" style="word-wrap: break-word;" /-->
```
Comment (guyguy2001) [inline src/views/feed/ActivityFeed/components/MessageItem.vue:22]: remove?

File: src/views/feed/ActivityFeed/components/MessageItem.vue [hunk 4]
```diff
@@ -34,25 +29,21 @@
...
 
-    @Prop({ default: false }) private first!: boolean;
+    @Prop() created!: string;
 
-    @Prop() private sender!: string;
-
-    @Prop() private article!: NewsItem;
-
-    @Prop() private type!: string;
+    @Prop() avatar!: string;
```
Comment (guyguy2001) [inline src/views/feed/ActivityFeed/components/MessageItem.vue:46]: `{required:true}`

File: src/views/feed/ActivityFeed/components/MessageThread.vue [hunk 5]
```diff
@@ -1,12 +1,12 @@
...
-      <MessageItem
+      <PrivateMessageItem
         v-for="(item, index) in messages"
         :key="item.created"
-        v-bind="item"
-        :article="article"
-        :first="index === 0"
+        :message="item"
+        :notification="notification"
+        :style="{paddingLeft: !(index === 0) && '50px'}"
```
Comment (guyguy2001) [inline src/views/feed/ActivityFeed/components/MessageThread.vue:9]: I'm confused by this line; If you put false does it not use the property? Also, why not `index !== 0`?
`&&` for with only one boolean is always confusing to me, so we should minimize all other complications that come alongside it IMO

File: src/views/feed/ActivityFeed/components/MessageThread.vue [hunk 6]
```diff
@@ -29,24 +29,24 @@
...
+  import PrivateMessageItem from './PrivateMessageItem.vue';
   import MessageCompose from './MessageCompose.vue';
 
   @Component({
-    components: { MessageItem, MessageCompose },
+    components: { PrivateMessageItem, MessageCompose },
   })
   export default class MessageThread extends Vue {
-    @Prop() private article!: NewsItem;
+    @Prop() private notification!: PrivateMessageNotificationItem;
```
Comment (guyguy2001) [inline src/views/feed/ActivityFeed/components/MessageThread.vue:40]: required

File: src/views/feed/ActivityFeed/components/PrivateMessageItem.vue [hunk 7]
```diff
@@ -0,0 +1,49 @@
+<template>
+  <MessageItem
+    :sender="message.sender"
+    :created="message.created"
+    :avatar="avatar"
+    :message="message.content"
+  >
+    <template> {{ senderName }} > {{ recieverName }}: </template>
```
Comment (guyguy2001) [inline src/views/feed/ActivityFeed/components/PrivateMessageItem.vue:8]: What is the `template` needed for?

File: src/views/feed/ActivityFeed/components/PrivateMessageItem.vue [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+  import Utils from '@/utils/utils';
+  import MessageItem from './MessageItem.vue';
+
+  @Component({
+    components: { MessageItem },
+  })
+  export default class PrivateMessageItem extends Vue {
+    @Prop() notification!: PrivateMessageNotificationItem;
+    
+    @Prop() message!: PrivateNotificationMessage;
```
Comment (guyguy2001) [inline src/views/feed/ActivityFeed/components/PrivateMessageItem.vue:23]: required

File: src/views/home/PlayerHome/components/activities/TutorialActivity.vue [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+
+  const PROGRESS_IMAGES = [
+    PROGRESS_IMAGE_0,
+    PROGRESS_IMAGE_1,
+    PROGRESS_IMAGE_2,
+    PROGRESS_IMAGE_3,
+    PROGRESS_IMAGE_4,
+    PROGRESS_IMAGE_5,
+    PROGRESS_IMAGE_6,
+  ];
```
Comment (guyguy2001) [inline src/views/home/PlayerHome/components/activities/TutorialActivity.vue:39]: code smell; Is there any way to do an array import (even if you specify the paths literally)?

File: src/views/home/PlayerHome/components/activities/TutorialActivity.vue [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+
+  @Component({
+    components: {
+      SwiperSlide,
+      BaseActivity,
+      QuestCard
+    }
+  })
+  export default class TutorialActivity extends Vue {
+    @Prop() tools!: RoadmapAchievement[];
```
Comment (guyguy2001) [inline src/views/home/PlayerHome/components/activities/TutorialActivity.vue:49]: required

File: src/views/news/NewsView/NewsView.vue [hunk 11]
```diff
@@ -64,16 +52,39 @@
...
+      this.created = res.news.created;
+      this.comments = res.comments;
     }
 
     private options: Option[] = [
       { value: 'categories', text: 'news-view:option-categories' },
       { value: 'announcements', text: 'news-view:option-announcements' },
       { value: 'blog', text: 'news-view:option-blog' },
-      { value: 'labs', text: 'news-view:option-labs' },
+      // { value: 'labs', text: 'news-view:option-labs' },
```
Comment (guyguy2001) [inline src/views/news/NewsView/NewsView.vue:87]: remove?

File: src/views/puzzles/PuzzleView/PuzzleView.vue [hunk 12]
```diff
@@ -1,8 +1,12 @@
 <template>
-  <EternaPage v-if="puzzle" :title="puzzle.title">
+  <!--
+    Tacking on the `&& puzzle` just to make our lives easier with typing,
+    as TS doesn't know that fetchState.firstFetchComplete correlates to it being filled out
```
Comment (guyguy2001) [inline src/views/puzzles/PuzzleView/PuzzleView.vue:4]: _give us reference to other members in type definitions_

File: src/components/PageLayout/EternaPage.vue [hunk 13]
```diff
@@ -62,7 +62,6 @@
 
     created() {
       this.unsubscribe = this.$vxm.mobile.$subscribe('showPageSidebar', payload => {
-        console.log('Eterna page: show sidebar was called');
```
[the remark this answers is not in the document]
Comment (luxaritas) [reply src/components/PageLayout/EternaPage.vue:65]: It actually is in our lint settings - but only for production (`process.env.NODE_ENV === 'production'`), so it doesn't spam the lint on save. However, when lint-staged/`vue-cli-service lint` gets run, it doesn't set NODE_ENV, so it never gets turned on. I briefly looked into how best to do that but couldn't find anything quickly - I'm going to leave it as is for now, but please let me know if you know how we should handle this

File: src/components/PageLayout/Navbar/Icons/NotificationIcon/NotificationIcon.vue [hunk 14]
```diff
@@ -0,0 +1,172 @@
...
+              @click="slotProp.hideDropdown.hide()"
+            />
+          </router-link>
+        </div>
+        <div class="border"></div>
+        <template v-for="item in notifications">
+          <NewsNotification v-if="isNewsItem(item)" :key="getCreated(item)" :article="item" />
+          <PrivateMessageNotification v-else-if="isPM(item)" :key="getCreated(item)" :pm="item" />
+          <CommentNotification v-else-if="isComment(item)" :key="getCreated(item)" :comment="item" />
+        </template>
```
[the remark this answers is not in the document]
Comment (luxaritas) [reply src/components/PageLayout/Navbar/Icons/NotificationIcon/NotificationIcon.vue:25]: Not saying you're wrong, but not sure exactly what your concern is here - let me know if you have additional thoughts

File: src/views/home/PlayerHome/components/activities/TutorialActivity.vue (hunk 9 above)
Comment (luxaritas) [reply src/views/home/PlayerHome/components/activities/TutorialActivity.vue:39]: No - but what should really happen is that it should be a dynamic SVG. Will file a separate issue

**Comment by luxaritas** [comment]
Thank you so much Guy - I believe I've gotten everything. Going to merge into dev so that hopefully we can deploy tomorrow

**Pull request merged**